pkg/scanners/template/scanner_template.go

```diff
@@ -0,0 +1,155 @@
+package template
+
+import (
+	"context"
+	"io"
+	"os"
+	"os/signal"
+	"syscall"
+
+	eraserv1alpha1 "github.com/Azure/eraser/api/v1alpha1"
+	"github.com/go-logr/logr"
+	"golang.org/x/sys/unix"
+
+	"github.com/Azure/eraser/pkg/metrics"
+	util "github.com/Azure/eraser/pkg/utils"
+	"go.opentelemetry.io/otel/metric/global"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// interface for custom scanners to communicate with Eraser
+type ImageProvider interface {
+	// receive list of all non-running, non-excluded images from collector container to process
+	ReceiveImages() []eraserv1alpha1.Image
+
+	// sends non-compliant images found to eraser container for removal
+	SendImages(nonCompliantImages, failedImages []eraserv1alpha1.Image)
+
+	// completes scanner communication process - required after custom scanning finishes
+	Finish()
+}
+
+type config struct {
+	ctx                    context.Context
+	log                    logr.Logger
+	deleteScanFailedImages bool
+	reportMetrics          bool
+}
+
+type ConfigFunc func(*config)
+
+func NewImageProvider(funcs ...ConfigFunc) ImageProvider {
+	// default config
+	cfg := &config{
+		ctx:                    context.Background(),
+		log:                    logf.Log.WithName("scanner"),
+		deleteScanFailedImages: true,
+		reportMetrics:          false,
+	}
+
+	// apply user config
+	for _, f := range funcs {
+		f(cfg)
+	}
+
+	return cfg
+}
+
+func (cfg *config) ReceiveImages() []eraserv1alpha1.Image {
+	var err error
+
+	if err := unix.Mkfifo(util.EraseCompleteScanPath, util.PipeMode); err != nil {
+		cfg.log.Error(err, "failed to create pipe", "pipeName", util.EraseCompleteScanPath)
+		os.Exit(1)
```

Comment:
In library code, it's pretty rude to exit or panic. You should let the caller decide what to do when there's an error, and that can be accomplished by "bubbling up" the error instead of taking decisive action when it happens. This goes for the other functions in the interface as well